core/src/main/java/org/apache/iceberg/BaseMetadataTable.java

```diff
@@ -186,4 +194,34 @@ public Transaction newTransaction() {
   public String toString() {
     return name();
   }
+
+  abstract Object writeReplace();
```

Comment:
This object has access to `table()`, `io()`, and `name()` already. Would it be easier to expose protected `metadataLocation()` and `metadataTableType()` methods instead of `writeReplace()` in each implementation? Then this method could be implemented here.

Reply:
Followed your recommendation. LGTM

The thing is that `metadataLocation()` is exactly the same for every implementation, but we need this if we do not want to relax the `private` restriction on `ops`. We might be further simplify the code if we move `ops` to `BaseMetadataTable`.

What do you think? Or it would be a bigger change which do no worth to do?

Reply:
We could make `ops` a protected method. Let's get this in and we can clean that up later.